### ISSUE TITLE
The performance of tieredImageNet

### ISSUE THREAD

**Issue by myyySun** [bbbdylan/unisiam#1]
We followed the settings of config/unisiam/tiered/r18.sh and re-trained the model with backbone resnet18 on tieredImageNet, and its performance as follows,

classifier: LR, power_norm: True, 5-way 1-shot acc: 63.24+0.40
classifier: LR, power_norm: True, 5-way 5-shot acc: 81.14+0.31

It severe below the paper, how can I improve the performance? Can you release the guide of how to train the model on tieredImageNet?

Thank you very much.

**Comment by lynlynlyn** [comment]
Please provide more details, such as the environment, GPU type, and PyTorch version. I can try to help you solve the problem.

**Comment by myyySun** [comment]
Thank you very much for your reply.

I didn't use the knowledge distillation and set the learning rate to 0.05 and the batch size to 512.
I trained the model on RTX 3090*8. the PyTorch is version 1.12.1.

the environment is as follows:
numpy                     1.23.5          py310hd5efca6_0  
numpy-base                1.23.5          py310h8e6c178_0  
openh264                  2.1.1                h4ff587b_0  
openjpeg                  2.4.0                h3ad879b_0  
openssl                   1.1.1t               h7f8727e_0  
packaging                 22.0            py310h06a4308_0  
pandas                    1.5.3                    pypi_0    pypi
partd                     1.2.0              pyhd3eb1b0_1  
pcre                      8.45                 h295c915_0  
pillow                    9.3.0           py310h6a678d5_2  
pip                       22.3.1          py310h06a4308_0  
ply                       3.11            py310h06a4308_0  
pooch                     1.4.0              pyhd3eb1b0_0  
pycparser                 2.21               pyhd3eb1b0_0  
pyopenssl                 22.0.0             pyhd3eb1b0_0  
pyparsing                 3.0.9           py310h06a4308_0  
pyqt                      5.15.7          py310h6a678d5_1  
pyqt5-sip                 12.11.0                  pypi_0    pypi
pysocks                   1.7.1           py310h06a4308_0  
python                    3.10.9               h7a1cb2a_0  
python-dateutil           2.8.2              pyhd3eb1b0_0  
pytorch                   1.12.1          py3.10_cuda11.3_cudnn8.3.2_0    pytorch
pytorch-mutex             1.0                        cuda    pytorch
pytz                      2022.7.1                 pypi_0    pypi
pywavelets                1.4.1           py310h5eee18b_0  
pyyaml                    6.0             py310h5eee18b_1  
qt-main                   5.15.2               h327a75a_7  
qt-webengine              5.15.9               hd2b0992_4  
qtwebkit                  5.212                h4eab89a_4  
qt-webengine              5.15.9               hd2b0992_4  
qtwebkit                  5.212                h4eab89a_4  
readline                  8.2                  h5eee18b_0  
requests                  2.28.1          py310h06a4308_0  
scikit-image              0.19.3          py310h6a678d5_1  
scikit-learn              1.2.1                    pypi_0    pypi
scipy                     1.10.0                   pypi_0    pypi
setuptools                65.6.3          py310h06a4308_0  
sip                       6.6.2           py310h6a678d5_0  
six                       1.16.0             pyhd3eb1b0_1  
sklearn                   0.0.post1                pypi_0    pypi
snappy                    1.1.9                h295c915_0  
sqlite                    3.40.1               h5082296_0  
threadpoolctl             3.1.0                    pypi_0    pypi
tifffile                  2021.7.2           pyhd3eb1b0_2  
tk                        8.6.12               h1ccaba5_0  
toml                      0.10.2             pyhd3eb1b0_0  
toolz                     0.12.0          py310h06a4308_0  
torchaudio                0.12.1              py310_cu113    pytorch
torchvision               0.13.1              py310_cu113    pytorch
tornado                   6.2             py310h5eee18b_0  
tqdm                      4.64.1          py310h06a4308_0  
typing_extensions         4.4.0           py310h06a4308_0  
tzdata                    2022g                h04d1e81_0  
urllib3                   1.26.14         py310h06a4308_0  
wheel                     0.37.1             pyhd3eb1b0_0  
xz                        5.2.10               h5eee18b_1  
yaml                      0.2.5                h7b6447c_0  
zfp                       0.5.5                h295c915_6  
zlib                      1.2.13               h5eee18b_0  
zstd                      1.5.2                ha4553b6_0

Is there anything wrong with this?

**Comment by myyySun** [comment]
Thank you very much for your reply.

I didn't use the knowledge distillation and set the learning rate to 0.05 and the batch size to 512.
I trained the model on RTX 3090*8. the PyTorch is version 1.12.1.

Is there anything wrong with this?

**Comment by lynlynlyn** [comment]
Since this implementation is based on PyTorch DataParallel, the statistical parameters of BN are sensitive to the number of GPUs. Our experiments were conducted on 4*V100 (32GB) GPUs.

**Comment by lynlynlyn** [comment]
What is the training loss on the last epoch? The loss_pos and loss_neg on the last epoch should be ~-0.892 and ~0.002.

**Comment by myyySun** [comment]
Would it be better if I trained on 4*3090? Or do I have to train on v100?

The training loss on the last epoch is as follows:
Train: [199][870/877]   BT 0.483 (0.530)        DT 0.000 (0.025)        loss -0.934 (-0.938)    loss_pos -0.934 (-0.938)       loss_neg 0.001 (0.001)  std 0.013 (0.013)
Is this normal?

**Comment by lynlynlyn** [comment]
I personally think that there shouldn't be a big difference between 3090 and V100. Your pos loss is too low.

**Comment by myyySun** [comment]
Why is the pos loss so low? How can I fix it?

**Comment by lynlynlyn** [comment]
It is recommended to run our code on 4 GPU.

**Comment by myyySun** [comment]
I will try the way you recommended. Thank you very much for your reply!

**Comment by myyySun** [comment]
Sorry to bother you again.

I re-trained it on 4*3090 again. But the performance did not improve. It‘s as follows,
classifier: LR, power_norm: True, 5-way 1-shot acc: 63.57+0.40
classifier: LR, power_norm: True, 5-way 5-shot acc: 80.77+0.31

And the loss of the last epoch is as follows,
Train: [199][870/877]	BT 0.412 (0.604)	DT 0.000 (0.184)	loss -0.899 (-0.894)	loss_pos -0.899 (-0.894)	loss_neg 0.001 (0.001)	std 0.009 (0.009)

The pos loss seems to be normal. What else might be wrong with it?

**Comment by lynlynlyn** [comment]
Based on the current information, I have difficulty pointing out where the problem lies. We re-verified this PyTorch re-implementation before releasing it. The setup (ResNet-18, tieredImageNet) has been tested in two PyTorch environments:(1) PyTorch-1.4.0, python3.6, cuda10.1; (2) PyTorch-1.8.0, python3.7, cuda10.2.  Both of them run on 4*V100 (32G).

The results and losses for the 1st environment (pytorch1.4.0) are
classifier: LR, power_norm: True, 5-way 1-shot acc: 64.81+0.40
classifier: LR, power_norm: True, 5-way 5-shot acc: 81.70+0.31
Train: [199][870/877] BT 0.542 (0.764) DT 0.000 (0.194) loss -0.899 (-0.891) loss_pos -0.899 (-0.892) loss_neg 0.002 (0.002) std 0.014 (0.014)

The results and losses for the 2nd environment (pytorch1.8.0) are
classifier: LR, power_norm: True, 5-way 1-shot acc: 65.45+0.39
classifier: LR, power_norm: True, 5-way 5-shot acc: 82.66+0.30
Train: [200][870/877] BT 0.402 (0.646) DT 0.000 (0.166) loss -0.858 (-0.870) loss_pos -0.858 (-0.870) loss_neg 0.000 (0.000) std 0.021 (0.021)

The difference in epochs is due to changes in some code (of print) between the two experiments, but there is essentially no difference.

If possible, I suggest conductingthe experiment in the pytorch-1.8.0 environment.

**Comment by myyySun** [comment]
ok, I wil try it. Thanks for your suggestion!

**Comment by myyySun** [comment]
I re-trained the model with PyTorch-1.8.0, python3.8, cuda11.1 on 4*3090 (24G). But the accuracy is even lower than previous. 

The results and losses are
classifier: LR, power_norm: True, 5-way 1-shot acc: 61.88+0.42
classifier: LR, power_norm: True, 5-way 5-shot acc: 78.06+0.33
Train: [199][870/877]	BT 0.531 (0.596)	DT 0.000 (0.012)	loss -0.927 (-0.925)	loss_pos -0.928 (-0.926)	loss_neg 0.014 (0.012)	std 0.012 (0.012)

It can be seen that the pos loss is too low and the neg loss is too large.

I would like to confirm that the initial learning rate for training is 0.05 and the batch size is 512? It‘s different in the paper and in config/unisiam/tiered/r18.sh.
Besides that, is the data augmentation strategy using stronger data augmentation or the default strategy?

**Comment by lynlynlyn** [comment]
The learning rate in config is the base learning rate. With the linear lr scaling, the real learning rate (reported in the paper) is 0.05*512/256=0.1. The data augmentation strategy is stronger data augmentation. 

Maybe you can set a fixed rand seed and comment out this line "cudnn.benchmark = True". I'm wondering if it's due to randomness.

**Comment by myyySun** [comment]
I set the random seed to 123 and commented out this line of code. The result is as follows:
classifier: LR, power_norm: True, 5-way 1-shot acc: 63.20+0.40
classifier: LR, power_norm: True, 5-way 5-shot acc: 80.87+0.31
Train: [199][870/877]	BT 0.475 (0.518)	DT 0.000 (0.013)	loss -0.869 (-0.874)	loss_pos -0.869 (-0.874)	loss_neg 0.000 (0.000)	std 0.022 (0.022)

This result is still somewhat lower than the one reported in the paper.

**Comment by lynlynlyn** [comment]
Which environment is this run in? To better solve the problem, I contacted you directly email.

**Comment by myyySun** [comment]
PyTorch-1.8.0, python3.8, cuda11.1. It run on 4*3090 (24G).

**Comment by myyySun** [comment]
After talking to the author, we eventually found that the problem was caused by the dataset. The size of the tieredImageNet I used is 84, while the author used the original image for ImageNet, which is 224.
According to the description of the tieredImageNet in ReadMe in this project github, I regenerated the dataset and retrained the models with two versions of PyTorch on 4*3090. The final result is as follows:

(PyTorch 1.8.0):
Train: [199][870/877]	BT 2.112 (0.828)	DT 1.647 (0.234)	loss -0.907 (-0.902)	loss_pos -0.907 (-0.903)	loss_neg 0.003 (0.004)	std 0.010 (0.009)
classifier: LR, power_norm: True, 5-way 1-shot acc: 64.84+0.40
classifier: LR, power_norm: True, 5-way 5-shot acc: 81.72+0.30

(PyTorch 1.12.0):
Train: [199][870/877]	BT 0.421 (0.843)	DT 0.000 (0.393)	loss -0.882 (-0.868)	loss_pos -0.882 (-0.869)	loss_neg 0.000 (0.000)	std 0.021 (0.021)
classifier: LR, power_norm: True, 5-way 1-shot acc: 64.74+0.41
classifier: LR, power_norm: True, 5-way 5-shot acc: 82.26+0.30

So the PyTorch version and the number of GPUs may not lead to a drop in performance. The image size of the dataset is important.
